lib/spack/llnl/util/symlink.py

```diff
@@ -2,77 +2,187 @@
 # Spack Project Developers. See the top-level COPYRIGHT file for details.
 #
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
-import errno
 import os
+import re
 import shutil
+import subprocess
 import sys
 import tempfile
-from os.path import exists, join
 
-from llnl.util import lang
+from llnl.util import lang, tty
+
+from spack.error import SpackError
+from spack.util.path import system_path_filter
 
 if sys.platform == "win32":
     from win32file import CreateHardLink
 
+is_windows = sys.platform == "win32"
+
 
-def symlink(real_path, link_path):
+def symlink(source_path: str, link_path: str, allow_broken_symlinks: bool = not is_windows):
     """
-    Create a symbolic link.
+    Create a link.
+
+    On non-Windows and Windows with System Administrator
+    privleges this will be a normal symbolic link via
+    os.symlink.
+
+    On Windows without privledges the link will be a
+    junction for a directory and a hardlink for a file.
+    On Windows the various link types are:
+
+    Symbolic Link: A link to a file or directory on the
+    same or different volume (drive letter) or even to
+    a remote file or directory (using UNC in its path).
+    Need System Administrator privileges to make these.
 
-    On Windows, use junctions if os.symlink fails.
+    Hard Link: A link to a file on the same volume (drive
+    letter) only. Every file (file's data) has at least 1
+    hard link (file's name). But when this method creates
+    a new hard link there will be 2. Deleting all hard
+    links effectively deletes the file. Don't need System
+    Administrator privileges.
+
+    Junction: A link to a directory on the same or different
+    volume (drive letter) but not to a remote directory. Don't
+    need System Administrator privileges.
+
+    Parameters:
+        source_path (str): The real file or directory that the link points to.
+            Must be absolute OR relative to the link.
+        link_path (str): The path where the link will exist.
+        allow_broken_symlinks (bool): On Linux or Mac, don't raise an exception if the source_path
+            doesn't exist. This will still raise an exception on Windows.
     """
-    if sys.platform != "win32":
-        os.symlink(real_path, link_path)
-    elif _win32_can_symlink():
-        # Windows requires target_is_directory=True when the target is a dir.
-        os.symlink(real_path, link_path, target_is_directory=os.path.isdir(real_path))
+    source_path = os.path.normpath(source_path)
+    win_source_path = source_path
+    link_path = os.path.normpath(link_path)
+
+    # Never allow broken links on Windows.
+    if sys.platform == "win32" and allow_broken_symlinks:
+        raise ValueError("allow_broken_symlinks parameter cannot be True on Windows.")
+
+    # Perform basic checks to make sure symlinking will succeed
+    if os.path.lexists(link_path):
```

Comment:
Can you please not do this? It's not atomic anyways. Can't you do everything in a tell-dont-ask way? Just make the symlink and catch the relevant error. I really don't want to see this overhead on unix.

Reply:
I'll put this inside of an `if not allow_broken_symlinks:` block, but this **HAS** to be done on Windows, there is no option. The link and target paths have to be resolved prior to trying to make a link. `os.symlink` does this automatically, but hard links and junctions just don't work that way.

Reply:
I think what @haampie is saying is that even if you have

```
        if os.path.lexists(link_path):
            raise SymlinkError(f"Link path ({link_path}) already exists. Cannot create link.")
```

that right after this check, some other process could create this link file, and that you could get the error anyway (this is one of the common examples of the better-to-ask-forgiveness-than-permission paradigm that is encouraged in Python programming).

I will say @haampie my understanding is the goal is to achieve a common exception format, and @jamessmillie FWIW a more-robust way to achieve that is to do the actual symlink/junction creation in a try/except block:

```
try:
    _windows_create_link(win_source_path, link_path)
except OSError as e:
    if this_is_a_file_already_exists_error_according_to_examination_of_the_exception:
        raise SymlinkError(...) from e
# This would have to be done for the `os.symlink` call as well, and perhaps you may
# decide that is too much effort to achieve a common exception format.
```

that being said, I'd prefer this be done in a follow-up: I assumed that most calls to `symlink.symlink` would occur for paths inside of installation prefixes, which either should be coordinated properly by prefix locks (not currently done on Windows, but that's a side note). There *are* some cases where multiple Spacks might simultaneously symlink into a common directory (xcode tools for `%apple_clang`) but AFAIK quite rare.